File: backend/utils/github_input.py
from urllib.parse import urlparse


def normalize_github_username(raw: str) -> str:
    """
    Accept a github username ot profile URL and return a clean username.

    Examples:
        "hliran2: -> "hliran2"
        "@hliran2" -> "hliran2"
        "https://github.com/hliran2" -> "hliran2"
        "https://github.com/hliran2/" -> "hliran2"
        "github.com/hliran2" -> "hliran2"
    """
    value = raw.strip()

    if not value:
        raise ValueError("Username cannot be empty")

    if value.startswith("@"):
        value = value[1:]

    if "github.com" in value.lower():
        if not value.startswith(("https://", "http://")):
            value = f"https://{value}"
        parsed_url = urlparse(value)
        path_parts = [part for part in parsed_url.path.strip("/").split("/") if part]

        if not path_parts:
            raise ValueError("No username found in GitHub URL")

        username = path_parts[0]
    else:
        username = value

    username = username.strip().strip("/")

    if not username:
        raise ValueError("Username cannot be empty")

    return username.lower()

File: frontend/src/utils/githubInput.ts
export function normalizeGitHubUsername(raw: string): string {
    let value = raw.trim()
  
    if (!value) {
      throw new Error('Username cannot be empty')
    }
  
    if (value.startsWith('@')) {
      value = value.slice(1).trim()
    }
  
    if (value.toLowerCase().includes('github.com')) {
      if (!value.startsWith('http://') && !value.startsWith('https://')) {
        value = `https://${value}`
      }
  
      const url = new URL(value)
      const pathParts = url.pathname
        .split('/')
        .map((part) => part.trim())
        .filter(Boolean)
  
      if (pathParts.length === 0) {
        throw new Error('No username found in GitHub URL')
      }
  
      value = pathParts[0]
    }
  
    value = value.trim().replace(/^\/+|\/+$/g, '')
  
    if (!value) {
      throw new Error('Username cannot be empty')
    }
  
    return value.toLowerCase()
  }